File: src/jira_dc_mcp/tools/fields.py
"""Field and field configuration tools."""

from __future__ import annotations

import json
import logging

from ..client import JiraClient, bounded_gather

logger = logging.getLogger(__name__)


async def list_fields(client: JiraClient, custom_only: bool = False, field_ids: list[str] | None = None) -> str:
    """List all fields (system + custom), optionally filtered to custom only or by specific IDs."""
    fields = await client.list_fields()
    if field_ids:
        id_set = set(field_ids)
        fields = [f for f in fields if f.get("id") in id_set]
    elif custom_only:
        fields = [f for f in fields if f.get("custom", False)]

    result = [
        {
            "id": f.get("id"),
            "name": f.get("name"),
            "custom": f.get("custom", False),
            "type": f.get("schema", {}).get("type") if f.get("schema") else None,
            "customType": f.get("schema", {}).get("custom") if f.get("schema") else None,
            "searchable": f.get("searchable"),
            "clauseNames": f.get("clauseNames", []),
        }
        for f in fields
    ]
    return json.dumps(result, indent=2)


async def get_field_configuration(client: JiraClient, fc_id: int) -> str:
    """Get field configuration items — required, hidden, renderer, description per field.

    Served by the jiraMcpFieldConfigurations ScriptRunner endpoint; the native
    /rest/api/2/fieldconfiguration endpoint 404s on Jira DC.
    """
    fc_list = await client.list_field_configurations()
    fc = next((c for c in fc_list if c.get("id") == fc_id), None)
    if fc is None:
        return json.dumps({"error": f"Field configuration {fc_id} not found"})

    result = {
        "fieldConfigurationId": fc.get("id"),
        "name": fc.get("name"),
        "description": fc.get("description", ""),
        "isDefault": fc.get("isDefault", False),
        "fields": [
            {
                "fieldId": item.get("fieldId"),
                "fieldName": item.get("fieldName"),
                "description": item.get("description", ""),
                "isRequired": item.get("isRequired", False),
                "isHidden": item.get("isHidden", False),
                "renderer": item.get("rendererType"),
            }
            for item in fc.get("fields", [])
        ],
    }
    return json.dumps(result, indent=2)


async def get_field_configuration_scheme(client: JiraClient, scheme_id: int) -> str:
    """Get field configuration scheme — maps issue types to field configurations.

    Served by the jiraMcpFieldConfigurationSchemes ScriptRunner endpoint; the
    native /rest/api/2/fieldconfigurationscheme endpoint 404s on Jira DC.
    """
    schemes = await client.list_field_configuration_schemes()
    scheme = next((s for s in schemes if s.get("id") == scheme_id), None)
    if scheme is None:
        return json.dumps({"error": f"Field configuration scheme {scheme_id} not found"})

    result = {
        "id": scheme.get("id"),
        "name": scheme.get("name"),
        "description": scheme.get("description", ""),
        "mappings": [
            {
                "issueTypeId": m.get("issueTypeId"),
                "issueTypeName": m.get("issueTypeName"),
                "fieldConfigurationId": m.get("fieldConfigId"),
                "fieldConfigurationName": m.get("fieldConfigName"),
            }
            for m in scheme.get("mappings", [])
        ],
        "projects": scheme.get("projects", []),
    }
    return json.dumps(result, indent=2)


async def find_field_usage(client: JiraClient, field_id: str) -> str:
    """Find where a field appears across all screens and field configurations."""
    # Search screens — fanned out with bounded concurrency (one fetch per screen).
    screens = await client.list_screens()

    async def _scan_screen(scr: dict) -> list[dict]:
        hits: list[dict] = []
        try:
            full = await client.get_screen_full(scr["id"])
        except Exception:
            return hits
        for tab in full.get("tabs", []):
            for f in tab.get("fields", []):
                if f.get("id") == field_id:
                    hits.append({
                        "screenId": scr["id"],
                        "screenName": scr.get("name"),
                        "tabId": tab.get("id"),
                        "tabName": tab.get("name"),
                    })
        return hits

    per_screen = await bounded_gather([_scan_screen(scr) for scr in screens])
    screen_hits = [hit for hits in per_screen for hit in hits]

    # Search field configurations (jiraMcpFieldConfigurations ScriptRunner endpoint;
    # each config embeds its field items, so no per-config calls are needed).
    fc_list = await client.list_field_configurations()
    fc_hits = []
    for fc in fc_list:
        for item in fc.get("fields", []):
            if item.get("fieldId") == field_id:
                fc_hits.append({
                    "fieldConfigId": fc.get("id"),
                    "fieldConfigName": fc.get("name"),
                    "isRequired": item.get("isRequired", False),
                    "isHidden": item.get("isHidden", False),
                })

    result = {
        "fieldId": field_id,
        "screens": screen_hits,
        "fieldConfigurations": fc_hits,
        "totalScreens": len(screen_hits),
        "totalFieldConfigs": len(fc_hits),
    }
    return json.dumps(result, indent=2)


async def get_createmeta_fields(
    client: JiraClient, project_key: str, issue_type_id: str
) -> str:
    """Get fields available on the CREATE screen for a project + issue type.

    Shows field name, required flag, allowed values (for select/radio/checkbox fields),
    and default values. Essential for understanding what an automation rule must set
    when creating an issue.
    """
    raw = await client.get_createmeta_fields(project_key, issue_type_id)
    result = []
    for f in raw:
        entry: dict = {
            "fieldId": f.get("fieldId"),
            "name": f.get("name"),
            "required": f.get("required", False),
            "schema": f.get("schema"),
            "hasDefaultValue": f.get("hasDefaultValue", False),
        }
        if f.get("allowedValues"):
            entry["allowedValues"] = [
                {
                    "id": v.get("id"),
                    "value": v.get("value") or v.get("name"),
                    "disabled": v.get("disabled", False),
                }
                for v in f["allowedValues"]
            ]
        if f.get("defaultValue"):
            entry["defaultValue"] = f["defaultValue"]
        result.append(entry)
    return json.dumps(result, indent=2)


async def list_custom_fields_usage(
    client: JiraClient,
    search: str | None = None,
    unused_only: bool = False,
    project_key: str | None = None,
    min_issues: int | None = None,
) -> str:
    """List custom fields with per-field usage stats (Jira DC 10 endpoint).

    Returns: id, name, type, searcherKey, projectsCount, projectKeys (resolved),
    isAllProjects, screensCount, issuesWithValue, lastValueUpdate (ISO + epoch).
    """
    from datetime import datetime, timezone

    raw = await client.list_custom_fields_usage()
    values = raw.get("values", [])

    projects = await client.list_projects()
    proj_map = {int(p["id"]): p["key"] for p in projects}

    target_pid = None
    if project_key:
        target_pid = next((pid for pid, k in proj_map.items() if k == project_key), None)
        if target_pid is None:
            return json.dumps({"error": f"Unknown project_key: {project_key}"})

    s = search.lower() if search else None
    result = []
    for f in values:
        if s and s not in (f.get("name") or "").lower():
            continue
        if unused_only and (f.get("issuesWithValue") or 0) != 0:
            continue
        if min_issues is not None and (f.get("issuesWithValue") or 0) < min_issues:
            continue
        if target_pid is not None and not f.get("isAllProjects") and target_pid not in (f.get("projectIds") or []):
            continue

        lvu = f.get("lastValueUpdate")
        result.append({
            "id": f.get("id"),
            "name": f.get("name"),
            "type": f.get("type"),
            "searcherKey": f.get("searcherKey"),
            "projectsCount": f.get("projectsCount"),
            "projectKeys": [proj_map.get(pid, f"#{pid}") for pid in (f.get("projectIds") or [])],
            "isAllProjects": f.get("isAllProjects", False),
            "screensCount": f.get("screensCount"),
            "issuesWithValue": f.get("issuesWithValue"),
            "lastValueUpdate": datetime.fromtimestamp(lvu / 1000, tz=timezone.utc).isoformat() if lvu else None,
            "lastValueUpdateEpoch": lvu,
        })

    result.sort(key=lambda r: -(r["issuesWithValue"] or 0))

    return json.dumps({
        "total": raw.get("total"),
        "returned": len(result),
        "fields": result,
    }, indent=2)


async def get_field_contexts(client: JiraClient, field_id: str) -> str:
    """Get custom field contexts — which projects and issue types the field is scoped to.

    Served by the jiraMcpCustomFieldContexts ScriptRunner endpoint (replaces the
    unsupported internal API /rest/internal/2/field/{id}/context).
    """
    fields = await client.list_custom_field_contexts(field_id)
    if not fields:
        return json.dumps({"error": f"No custom field context data for field {field_id}"})
    entry = fields[0]
    contexts = [
        {
            "id": ctx.get("id"),
            "name": ctx.get("name", ""),
            "description": ctx.get("description", ""),
            "isAllProjects": ctx.get("isGlobalProjects", False),
            "isAllIssueTypes": ctx.get("isAllIssueTypes", False),
            "projects": ctx.get("projects", []),
            "issueTypes": ctx.get("issueTypes", []),
        }
        for ctx in entry.get("contexts", [])
    ]
    return json.dumps({
        "fieldId": entry.get("fieldId"),
        "fieldName": entry.get("fieldName"),
        "fieldType": entry.get("fieldType"),
        "contexts": contexts,
    }, indent=2)

File: src/tools/fields.ts
/** Field and field configuration tools. */

import { z } from "zod";
import { boundedAll } from "../client.js";
import { dumps } from "../json.js";
import type { ToolDef } from "./types.js";

/**
 * Format an epoch-milliseconds timestamp the way Python's
 * `datetime.fromtimestamp(ms/1000, tz=utc).isoformat()` does:
 * `YYYY-MM-DDTHH:MM:SS[.ffffff]+00:00` (6-digit microseconds, omitted when 0).
 */
export function pythonIsoUtc(epochMs: number): string {
  const d = new Date(epochMs);
  const p = (n: number, w = 2) => String(n).padStart(w, "0");
  const base =
    `${p(d.getUTCFullYear(), 4)}-${p(d.getUTCMonth() + 1)}-${p(d.getUTCDate())}` +
    `T${p(d.getUTCHours())}:${p(d.getUTCMinutes())}:${p(d.getUTCSeconds())}`;
  const micros = d.getUTCMilliseconds() * 1000;
  const frac = micros === 0 ? "" : `.${p(micros, 6)}`;
  return `${base}${frac}+00:00`;
}

export const fieldTools: ToolDef[] = [
  {
    name: "list_fields",
    description:
      "List all fields (system + custom) with types, search clause names. " +
      "Set custom_only=true to see only custom fields. " +
      "Pass field_ids to look up specific fields by ID.",
    inputShape: {
      custom_only: z.boolean().default(false).describe("Only show custom fields"),
      field_ids: z
        .array(z.string())
        .optional()
        .describe(
          "List of specific field IDs to look up (e.g. ['customfield_12808', 'summary']). " +
            "When provided, returns only these fields.",
        ),
    },
    async handler({ client }, args) {
      let fields = await client.listFields();
      if (args.field_ids && args.field_ids.length > 0) {
        const idSet = new Set<string>(args.field_ids);
        fields = fields.filter((f: any) => idSet.has(f.id));
      } else if (args.custom_only) {
        fields = fields.filter((f: any) => f.custom ?? false);
      }
      return dumps(
        fields.map((f: any) => ({
          id: f.id,
          name: f.name,
          custom: f.custom ?? false,
          type: f.schema ? (f.schema.type ?? null) : null,
          customType: f.schema ? (f.schema.custom ?? null) : null,
          searchable: f.searchable,
          clauseNames: f.clauseNames ?? [],
        })),
      );
    },
  },

  {
    name: "list_custom_fields_usage",
    description:
      "List custom fields with usage stats from Jira DC 10's /customFields endpoint: " +
      "issuesWithValue (issue count), projectsCount/keys, screensCount, lastValueUpdate. " +
      "Filters: search (substring on name), unused_only (issuesWithValue=0), " +
      "project_key (limits to fields scoped to that project), min_issues (>= N issues). " +
      "Sorted by issue count desc. Useful for auditing dead custom fields.",
    inputShape: {
      search: z.string().optional().describe("Case-insensitive substring filter on field name"),
      unused_only: z
        .boolean()
        .default(false)
        .describe("Only fields with zero issues using them"),
      project_key: z
        .string()
        .optional()
        .describe(
          "Project key (e.g. FINJ) — limits to fields scoped to that project or isAllProjects=true",
        ),
      min_issues: z
        .coerce.number()
        .int()
        .optional()
        .describe("Only fields with issuesWithValue >= this"),
    },
    async handler({ client }, args) {
      const raw = await client.listCustomFieldsUsage();
      const values: any[] = raw.values ?? [];

      const projects = await client.listProjects();
      const projMap = new Map<number, string>();
      for (const p of projects) projMap.set(Number(p.id), p.key);

      let targetPid: number | null = null;
      if (args.project_key) {
        targetPid = null;
        for (const [pid, k] of projMap) {
          if (k === args.project_key) {
            targetPid = pid;
            break;
          }
        }
        if (targetPid === null) {
          return dumps({ error: `Unknown project_key: ${args.project_key}` });
        }
      }

      const s = args.search ? String(args.search).toLowerCase() : null;
      const result: any[] = [];
      for (const f of values) {
        if (s && !(f.name ?? "").toLowerCase().includes(s)) continue;
        if (args.unused_only && (f.issuesWithValue ?? 0) !== 0) continue;
        if (args.min_issues !== undefined && args.min_issues !== null && (f.issuesWithValue ?? 0) < args.min_issues) {
          continue;
        }
        if (
          targetPid !== null &&
          !f.isAllProjects &&
          !(f.projectIds ?? []).includes(targetPid)
        ) {
          continue;
        }

        const lvu = f.lastValueUpdate;
        result.push({
          id: f.id,
          name: f.name,
          type: f.type,
          searcherKey: f.searcherKey,
          projectsCount: f.projectsCount,
          projectKeys: (f.projectIds ?? []).map((pid: number) => projMap.get(pid) ?? `#${pid}`),
          isAllProjects: f.isAllProjects ?? false,
          screensCount: f.screensCount,
          issuesWithValue: f.issuesWithValue,
          lastValueUpdate: lvu ? pythonIsoUtc(lvu) : null,
          lastValueUpdateEpoch: lvu,
        });
      }

      result.sort((a, b) => (b.issuesWithValue || 0) - (a.issuesWithValue || 0));

      return dumps({ total: raw.total, returned: result.length, fields: result });
    },
  },

  {
    name: "get_field_configuration",
    description:
      "Get field configuration items — shows which fields are required, hidden, " +
      "their renderer and description. The 'rules' for fields.",
    inputShape: { fc_id: z.coerce.number().int().describe("Field configuration ID") },
    async handler({ client }, args) {
      const fcList = await client.listFieldConfigurations();
      const fc = fcList.find((c: any) => c.id === args.fc_id);
      if (fc === undefined) {
        return dumps({ error: `Field configuration ${args.fc_id} not found` });
      }
      return dumps({
        fieldConfigurationId: fc.id,
        name: fc.name,
        description: fc.description ?? "",
        isDefault: fc.isDefault ?? false,
        fields: (fc.fields ?? []).map((item: any) => ({
          fieldId: item.fieldId,
          fieldName: item.fieldName,
          description: item.description ?? "",
          isRequired: item.isRequired ?? false,
          isHidden: item.isHidden ?? false,
          renderer: item.rendererType,
        })),
      });
    },
  },

  {
    name: "get_field_configuration_scheme",
    description:
      "Get field configuration scheme — maps issue types to field configurations.",
    inputShape: {
      scheme_id: z.coerce.number().int().describe("Field configuration scheme ID"),
    },
    async handler({ client }, args) {
      const schemes = await client.listFieldConfigurationSchemes();
      const scheme = schemes.find((s: any) => s.id === args.scheme_id);
      if (scheme === undefined) {
        return dumps({ error: `Field configuration scheme ${args.scheme_id} not found` });
      }
      return dumps({
        id: scheme.id,
        name: scheme.name,
        description: scheme.description ?? "",
        mappings: (scheme.mappings ?? []).map((m: any) => ({
          issueTypeId: m.issueTypeId,
          issueTypeName: m.issueTypeName,
          fieldConfigurationId: m.fieldConfigId,
          fieldConfigurationName: m.fieldConfigName,
        })),
        projects: scheme.projects ?? [],
      });
    },
  },

  {
    name: "find_field_usage",
    description:
      "Find where a field appears across ALL screens. " +
      "Use to understand impact before modifying a field.",
    inputShape: {
      field_id: z.string().describe("Field ID (e.g. 'customfield_10001' or 'summary')"),
    },
    async handler({ client }, args) {
      const fieldId = args.field_id;
      const screens = await client.listScreens();

      const scanScreen = async (scr: any): Promise<any[]> => {
        const hits: any[] = [];
        let full: any;
        try {
          full = await client.getScreenFull(scr.id);
        } catch {
          return hits;
        }
        for (const tab of full.tabs ?? []) {
          for (const f of tab.fields ?? []) {
            if (f.id === fieldId) {
              hits.push({
                screenId: scr.id,
                screenName: scr.name,
                tabId: tab.id,
                tabName: tab.name,
              });
            }
          }
        }
        return hits;
      };

      const perScreen = await boundedAll(screens.map((scr: any) => () => scanScreen(scr)));
      const screenHits = perScreen.flat();

      const fcList = await client.listFieldConfigurations();
      const fcHits: any[] = [];
      for (const fc of fcList) {
        for (const item of fc.fields ?? []) {
          if (item.fieldId === fieldId) {
            fcHits.push({
              fieldConfigId: fc.id,
              fieldConfigName: fc.name,
              isRequired: item.isRequired ?? false,
              isHidden: item.isHidden ?? false,
            });
          }
        }
      }

      return dumps({
        fieldId,
        screens: screenHits,
        fieldConfigurations: fcHits,
        totalScreens: screenHits.length,
        totalFieldConfigs: fcHits.length,
      });
    },
  },

  {
    name: "get_createmeta_fields",
    description:
      "Get fields available on the CREATE screen for a project + issue type. " +
      "Shows field name, required flag, allowed values (for select/radio/checkbox fields), " +
      "and default values. Use to discover what values an automation rule must set.",
    inputShape: {
      project_key: z.string().describe("Jira project key (e.g. 'CL')"),
      issue_type_id: z.string().describe("Issue type ID (e.g. '13602')"),
    },
    async handler({ client }, args) {
      const raw = await client.getCreatemetaFields(args.project_key, args.issue_type_id);
      const result: any[] = [];
      for (const f of raw) {
        const entry: Record<string, unknown> = {
          fieldId: f.fieldId,
          name: f.name,
          required: f.required ?? false,
          schema: f.schema,
          hasDefaultValue: f.hasDefaultValue ?? false,
        };
        if (f.allowedValues && f.allowedValues.length > 0) {
          entry.allowedValues = f.allowedValues.map((v: any) => ({
            id: v.id,
            value: v.value || v.name,
            disabled: v.disabled ?? false,
          }));
        }
        if (f.defaultValue != null && Object.keys(f.defaultValue).length > 0) {
          entry.defaultValue = f.defaultValue;
        }
        result.push(entry);
      }
      return dumps(result);
    },
  },

  {
    name: "get_field_contexts",
    description:
      "Get custom field contexts — which projects and issue types the field is scoped to. " +
      "Uses an internal API (unsupported, may break on upgrades). " +
      "Shows allProjects/allIssueTypes flags and specific project/issue type lists.",
    inputShape: { field_id: z.string().describe("Field ID (e.g. 'customfield_10001')") },
    async handler({ client }, args) {
      const fields = await client.listCustomFieldContexts(args.field_id);
      if (fields.length === 0) {
        return dumps({ error: `No custom field context data for field ${args.field_id}` });
      }
      const entry = fields[0];
      return dumps({
        fieldId: entry.fieldId,
        fieldName: entry.fieldName,
        fieldType: entry.fieldType,
        contexts: (entry.contexts ?? []).map((ctx: any) => ({
          id: ctx.id,
          name: ctx.name ?? "",
          description: ctx.description ?? "",
          isAllProjects: ctx.isGlobalProjects ?? false,
          isAllIssueTypes: ctx.isAllIssueTypes ?? false,
          projects: ctx.projects ?? [],
          issueTypes: ctx.issueTypes ?? [],
        })),
      });
    },
  },
];
